src/server/builder.ts

```diff
@@ -0,0 +1,384 @@
+                const currentScriptInfos = this.project.getScriptInfos();
+                for (const scriptInfo of currentScriptInfos) {
+                    const fileInfo = this.getOrCreateFileInfo(scriptInfo.path);
+                    this.updateFileReferences(fileInfo);
+                }
+                this.forEachFileInfo(fileInfo => {
+                    if (!this.project.containsScriptInfo(fileInfo.scriptInfo)) {
+                        // This file was deleted from this project
+                        this.removeFileReferences(fileInfo);
+                        this.removeFileInfo(fileInfo.scriptInfo.path);
+                    }
+                });
+                this.projectVersionForDependencyGraph = this.project.getProjectVersion();
+            }
+        }
+
+        private updateFileReferences(fileInfo: ModuleBuilderFileInfo) {
+            // Only need to update if the content of the file changed.
+            if (fileInfo.scriptVersionForReferences === fileInfo.scriptInfo.getLatestVersion()) {
+                return;
+            }
+
+            const newReferences = this.getReferencedFileInfos(fileInfo);
+            const oldReferences = fileInfo.references;
+
+            let oldIndex = 0;
+            let newIndex = 0;
+            while (oldIndex < oldReferences.length && newIndex < newReferences.length) {
+                const oldReference = oldReferences[oldIndex];
+                const newReference = newReferences[newIndex];
+                const compare = ModuleBuilderFileInfo.compareFileInfos(oldReference, newReference);
+                if (compare < 0) {
+                    // New reference is greater then current reference. That means
+                    // the current reference doesn't exist anymore after parsing. So delete
+                    // references.
+                    oldReference.removeReferencedBy(fileInfo);
+                    oldIndex++;
+                }
+                else if (compare > 0) {
+                    // A new reference info. Add it.
+                    newReference.addReferencedBy(fileInfo);
+                    newIndex++;
+                }
+                else {
+                    // Equal. Go to next
+                    oldIndex++;
+                    newIndex++;
+                }
+            }
+            // Clean old references
+            for (let i = oldIndex; i < oldReferences.length; i++) {
+                oldReferences[i].removeReferencedBy(fileInfo);
+            }
+            // Update new references
+            for (let i = newIndex; i < newReferences.length; i++) {
+                newReferences[i].addReferencedBy(fileInfo);
+            }
+
+            fileInfo.references = newReferences;
+            fileInfo.scriptVersionForReferences = fileInfo.scriptInfo.getLatestVersion();
+        }
+
+        private removeFileReferences(fileInfo: ModuleBuilderFileInfo) {

```

Comment:
nit: can be made a method of `ModuleBuilderFileInfo` since it does not use anything from `ModuleBuilder`